File: dvbViewerController/src/main/java/org/dvbviewer/controller/ui/fragments/Remote.java
/*
 * Copyright © 2013 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.ui.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatSpinner;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.loader.app.LoaderManager.LoaderCallbacks;
import androidx.loader.content.Loader;
import androidx.viewpager.widget.ViewPager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.apache.commons.lang3.StringUtils;
import org.dvbviewer.controller.R;
import org.dvbviewer.controller.entities.DVBViewerPreferences;
import org.dvbviewer.controller.io.RecordingService;
import org.dvbviewer.controller.io.ServerRequest;
import org.dvbviewer.controller.ui.base.AsyncLoader;
import org.dvbviewer.controller.utils.ServerConsts;
import org.dvbviewer.controller.utils.UIUtils;

import java.lang.reflect.Type;
import java.text.MessageFormat;
import java.util.LinkedList;
import java.util.List;

/**
 * The Class Remote.
 *
 * @author RayBa
 * @date 07.04.2013
 */
public class Remote extends Fragment implements LoaderCallbacks<List<String>>, RemoteControl.OnRemoteButtonClickListener {

    private Toolbar mToolbar;
    private ArrayAdapter mSpinnerAdapter;
    private AppCompatSpinner mClientSpinner;
    private int spinnerPosition;
    private static final String KEY_SPINNER_POS = "spinnerPosition";
    private DVBViewerPreferences prefs;
    private final Gson gson = new Gson();
    private final Type type = new TypeToken<List<String>>() {
    }.getType();
    private ViewPager mPager;
    private OnTargetsChangedListener onTargetsChangedListener;

    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#onCreate(android.os.Bundle)
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#onActivityCreated(android.os.Bundle)
     */
    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (!UIUtils.isTablet(getContext())) {
            ((AppCompatActivity) getActivity()).setSupportActionBar(mToolbar);
        }
        initActionBar();
        prefs = new DVBViewerPreferences(getContext());
        if (savedInstanceState != null) {
            spinnerPosition = savedInstanceState.getInt(KEY_SPINNER_POS, 0);
        }
        mPager.setAdapter(new PagerAdapter(getChildFragmentManager()));
    }

    private void initActionBar() {
        ActionBar ab = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (ab != null){
            ab.setDisplayHomeAsUpEnabled(true);
        }
        mToolbar.setVisibility(onTargetsChangedListener == null ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnTargetsChangedListener){
            onTargetsChangedListener = (OnTargetsChangedListener) context;
        }
    }

    /* (non-Javadoc)
         * @see android.support.v4.app.Fragment#onCreateView(android.view.LayoutInflater, android.view.ViewGroup, android.os.Bundle)
         */
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_remote, container, false);
        mToolbar = v.findViewById(R.id.toolbar);

        // Set an OnMenuItemClickListener to handle menu item clicks
        mToolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                // Handle the menu item
                return true;
            }
        });

        mToolbar.setTitle(R.string.remote);
        mClientSpinner = v.findViewById(R.id.clientSpinner);
        mClientSpinner.setVisibility(View.GONE);
        mClientSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                String selectedClient = (String) mSpinnerAdapter.getItem(position);
                prefs.getPrefs().edit().putString(DVBViewerPreferences.KEY_SELECTED_CLIENT, selectedClient).commit();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
        return v;
    }

    /*
	 * (non-Javadoc)
	 *
	 * @see android.support.v4.app.Fragment#onViewCreated(android.view.View,
	 * android.os.Bundle)
	 */
    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPager = (ViewPager) view.findViewById(R.id.pager);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putInt(KEY_SPINNER_POS, mClientSpinner.getSelectedItemPosition());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onResume() {
        super.onResume();
        getLoaderManager().initLoader(0, null, this);
    }


    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#toString()
     */
    @Override
    public String toString() {
        return "Remote";
    }



    @Override
    public Loader<List<String>> onCreateLoader(int id, Bundle args) {
        AsyncLoader<List<String>> loader = new AsyncLoader<List<String>>(getContext()) {

            @Override
            public List<String> loadInBackground() {
                List<String> result = null;
                try {
                    result = getDVBViewerClients();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return result;
            }
        };
        return loader;
    }

    @Override
    public void onLoadFinished(Loader<List<String>> loader, List<String> data) {
        if (onTargetsChangedListener != null) {
            onTargetsChangedListener.targetsChanged(getString(R.string.remote), data);
        } else if (data != null && !data.isEmpty()) {
            String[] arr = new String[data.size()];
            mSpinnerAdapter = new ArrayAdapter<>(getContext(), R.layout.support_simple_spinner_dropdown_item, data.toArray(arr));
            mSpinnerAdapter.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item);
            mClientSpinner.setAdapter(mSpinnerAdapter);
            String activeClient = prefs.getString(DVBViewerPreferences.KEY_SELECTED_CLIENT);
            int index = data.indexOf(activeClient);
            spinnerPosition = index > Spinner.INVALID_POSITION ? index : Spinner.INVALID_POSITION;
            mClientSpinner.setSelection(spinnerPosition);
            mClientSpinner.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onLoaderReset(Loader<List<String>> loader) {
        loader.reset();
    }

    private List<String> getDVBViewerClients() {
        List<String> result = new LinkedList<>();
        String jsonClients = RecordingService.INSTANCE.getDvbViewerTargets();
        if (StringUtils.isNotBlank(jsonClients)) {
            result = gson.fromJson(jsonClients, type);
            SharedPreferences.Editor prefEditor = prefs.getPrefs().edit();
            prefEditor.putString(DVBViewerPreferences.KEY_RS_CLIENTS, jsonClients);
            prefEditor.apply();
        } else {
            final String prefValue = prefs.getPrefs().getString(DVBViewerPreferences.KEY_RS_CLIENTS, "");
            if (StringUtils.isNotBlank(prefValue)) {
                result = gson.fromJson(jsonClients, type);
            }
        }
        return result;
    }

    @Override
    public void OnRemoteButtonClick(String action) {
        final Object target = onTargetsChangedListener != null ? onTargetsChangedListener.getSelectedTarget() : mClientSpinner.getSelectedItem();
        if (target == null) {
            return;
        }
        final String request = MessageFormat.format(ServerConsts.REC_SERVICE_URL + ServerConsts.URL_SEND_COMMAND, target, action);
        ServerRequest.DVBViewerCommand httpCommand = new ServerRequest.DVBViewerCommand(getContext(), request);
        Thread executionThread = new Thread(httpCommand);
        executionThread.start();
    }

    /**
     * The Class PagerAdapter.
     *
     * @author RayBa
     * @date 07.04.2013
     */
    class PagerAdapter extends FragmentPagerAdapter {

        /**
         * Instantiates a new pager adapter.
         *
         * @param fm the fm
         * @author RayBa
         * @date 07.04.2013
         */
        public PagerAdapter(FragmentManager fm) {
            super(fm);
        }

        /*
         * (non-Javadoc)
         *
         * @see android.support.v4.app.FragmentPagerAdapter#getItem(int)
         */
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    RemoteControl ctl = (RemoteControl) Fragment.instantiate(getContext(), RemoteControl.class.getName());
                    return ctl;
                case 1:
                    RemoteNumbers numbers = (RemoteNumbers) Fragment.instantiate(getContext(), RemoteNumbers.class.getName());
                    return numbers;
                default:
                    return null;
            }
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return getString(R.string.remote_control);
                case 1:
                    return getString(R.string.remote_numbers);
                default:
                    return "";
            }
        }

        /*
                 * (non-Javadoc)
                 *
                 * @see android.support.v4.view.PagerAdapter#getCount()
                 */
        @Override
        public int getCount() {
            return 2;
        }

    }

    public interface OnTargetsChangedListener {

        void targetsChanged(String title, List<String> tragets);

        Object getSelectedTarget();

    }


}

File: dvbViewerController/src/main/java/org/dvbviewer/controller/ui/fragments/Remote.kt
/*
 * Copyright © 2013 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.ui.fragments

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatSpinner
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager.widget.ViewPager
import okhttp3.ResponseBody
import org.apache.commons.collections4.CollectionUtils
import org.dvbviewer.controller.R
import org.dvbviewer.controller.data.api.ApiResponse
import org.dvbviewer.controller.data.api.ApiStatus
import org.dvbviewer.controller.data.entities.DVBTarget
import org.dvbviewer.controller.data.entities.DVBViewerPreferences
import org.dvbviewer.controller.data.remote.RemoteRepository
import org.dvbviewer.controller.data.remote.RemoteViewModel
import org.dvbviewer.controller.data.remote.RemoteViewModelFactory
import org.dvbviewer.controller.ui.base.AbstractRemote
import org.dvbviewer.controller.ui.base.BaseFragment
import org.dvbviewer.controller.utils.UIUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

/**
 * The Class Remote.
 *
 * @author RayBa
 * @date 07.04.2013
 */
class Remote : BaseFragment(), AbstractRemote.OnRemoteButtonClickListener {

    private var mToolbar: Toolbar? = null
    private var mSpinnerAdapter: ArrayAdapter<*>? = null
    private var mClientSpinner: AppCompatSpinner? = null
    private var spinnerPosition: Int = 0
    private var onTargetsChangedListener: OnTargetsChangedListener? = null
    private lateinit var prefs: DVBViewerPreferences
    private lateinit var mPager: ViewPager
    private lateinit var repository: RemoteRepository
    private lateinit var viewModel: RemoteViewModel

    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#onCreate(android.os.Bundle)
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        repository = RemoteRepository(getDmsInterface())
        prefs = DVBViewerPreferences(activity?.applicationContext)
        val vFac = RemoteViewModelFactory(repository, prefs)
        viewModel = ViewModelProvider(this, vFac)
                .get(RemoteViewModel::class.java)
    }

    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#onActivityCreated(android.os.Bundle)
     */
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (!UIUtils.isTablet(context!!)) {
            (activity as AppCompatActivity).setSupportActionBar(mToolbar)
        }
        initActionBar()
        if (savedInstanceState != null) {
            spinnerPosition = savedInstanceState.getInt(KEY_SPINNER_POS, 0)
        }
        mPager.adapter = PagerAdapter(childFragmentManager)
        val targetObserver = Observer<ApiResponse<List<DVBTarget>>> { response -> onTargetsLoaded(response) }
        viewModel.getTargets().observe(this, targetObserver)
    }

    private fun onTargetsLoaded(observable: ApiResponse<List<DVBTarget>>?) {
        if (observable?.status == ApiStatus.SUCCESS) {
            if (onTargetsChangedListener != null) {
                onTargetsChangedListener?.targetsChanged(getString(R.string.remote), observable.data)
            } else if (CollectionUtils.isNotEmpty(observable.data)) {
                val clients = LinkedList<String>()
                observable.data?.map { it.name }?.let { clients.addAll(it) }
                mSpinnerAdapter = ArrayAdapter(context!!, R.layout.support_simple_spinner_dropdown_item, clients.toTypedArray())
                mSpinnerAdapter!!.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item)
                mClientSpinner?.adapter = mSpinnerAdapter
                val activeClient = prefs.getString(DVBViewerPreferences.KEY_SELECTED_CLIENT)
                val index = clients.indexOf(activeClient)
                spinnerPosition = if (index > Spinner.INVALID_POSITION) index else Spinner.INVALID_POSITION
                mClientSpinner?.setSelection(spinnerPosition)
                mClientSpinner?.visibility = View.VISIBLE
            }
        } else if (observable?.status == ApiStatus.ERROR) {
            catchException(TimerList.TAG, observable.e)
        }
    }


    private fun initActionBar() {
        val ab = (activity as AppCompatActivity).supportActionBar
        ab?.setDisplayHomeAsUpEnabled(true)
        mToolbar?.visibility = if (onTargetsChangedListener == null) View.VISIBLE else View.GONE
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnTargetsChangedListener) {
            onTargetsChangedListener = context
        }
    }

    /* (non-Javadoc)
         * @see android.support.v4.app.Fragment#onCreateView(android.view.LayoutInflater, android.view.ViewGroup, android.os.Bundle)
         */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_remote, container, false)
        mToolbar = v.findViewById(R.id.toolbar)

        // Set an OnMenuItemClickListener to handle menu item clicks
        mToolbar?.setOnMenuItemClickListener {
            // Handle the menu item
            true
        }

        mToolbar?.setTitle(R.string.remote)
        mClientSpinner = v.findViewById(R.id.clientSpinner)
        mClientSpinner?.visibility = View.GONE
        mClientSpinner?.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                val selectedClient = mSpinnerAdapter!!.getItem(position) as String?
                prefs.prefs.edit()
                        .putString(DVBViewerPreferences.KEY_SELECTED_CLIENT, selectedClient)
                        .apply()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }
        return v
    }

    /*
	 * (non-Javadoc)
	 *
	 * @see android.support.v4.app.Fragment#onViewCreated(android.view.View,
	 * android.os.Bundle)
	 */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mPager = view.findViewById<View>(R.id.pager) as ViewPager
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mClientSpinner?.selectedItemPosition?.let { outState.putInt(KEY_SPINNER_POS, it) }
        super.onSaveInstanceState(outState)
    }

    /* (non-Javadoc)
     * @see android.support.v4.app.Fragment#toString()
     */
    override fun toString(): String {
        return "Remote"
    }

    override fun OnRemoteButtonClick(action: String) {
        val target = if (onTargetsChangedListener != null) onTargetsChangedListener!!.selectedTarget else mClientSpinner!!.selectedItem
        if (target == null) {
            sendMessage(R.string.no_remote_target)
            return
        }
        repository.sendCommand(target.toString(), action).enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                Log.i(Remote::class.java.simpleName, "Send command to target")
            }

            override fun onFailure(call: Call<ResponseBody>, throwable: Throwable) {
                catchException(Remote::class.java.simpleName, throwable)
            }
        })
    }

    internal inner class PagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        /*
         * (non-Javadoc)
         *
         * @see android.support.v4.app.FragmentPagerAdapter#getItem(int)
         */
        override fun getItem(position: Int): Fragment {
            return when (position) {
                0 -> {
                    fragmentManager!!.fragmentFactory.instantiate(javaClass.classLoader!!, RemoteControl::class.java.name)
                }
                else -> {
                    fragmentManager!!.fragmentFactory.instantiate(javaClass.classLoader!!, RemoteNumbers::class.java.name)
                }
            }
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return when (position) {
                0 -> getString(R.string.remote_control)
                1 -> getString(R.string.remote_numbers)
                else -> ""
            }
        }

        /*
                 * (non-Javadoc)
                 *
                 * @see android.support.v4.view.PagerAdapter#getCount()
                 */
        override fun getCount(): Int {
            return 2
        }

    }

    interface OnTargetsChangedListener {

        var selectedTarget: DVBTarget?

        fun targetsChanged(title: String, tragets: List<DVBTarget>?)

    }

    companion object {
        private const val KEY_SPINNER_POS = "spinnerPosition"
    }


}
